autopilot/contractor.go

```diff
@@ -1548,30 +1587,163 @@ func (c *contractor) formContract(ctx context.Context, w Worker, host hostdb.Hos
 	return formedContract, true, nil
 }
 
-func refreshPriceTable(ctx context.Context, w Worker, host *hostdb.Host) error {
-	// return early if the host's pricetable is not expired yet
-	if !host.PriceTable.Expiry.IsZero() && time.Now().After(host.PriceTable.Expiry) {
-		return nil
+func (c *contractor) tryPerformPruning(ctx context.Context, wp *workerPool) {
+	c.mu.Lock()
+	if c.pruning || c.ap.isStopped() {
+		c.mu.Unlock()
+		return
 	}
+	c.pruning = true
+	c.pruningLastStart = time.Now()
+	c.mu.Unlock()
 
-	// scan the host if it hasn't been successfully scanned before, which
-	// can occur when contracts are added manually to the bus or database
-	if !host.Scanned {
-		scan, err := w.RHPScan(ctx, host.PublicKey, host.NetAddress, timeoutHostScan)
-		if err != nil {
-			return fmt.Errorf("failed to scan host %v: %w", host.PublicKey, err)
+	c.ap.wg.Add(1)
+	go func() {
+		defer c.ap.wg.Done()
+		c.performContractPruning(wp)
+		c.mu.Lock()
+		c.pruning = false
+		c.mu.Unlock()
+	}()
+}
+
+func (c *contractor) performContractPruning(wp *workerPool) {
+	c.logger.Info("performing contract pruning")
+
+	// fetch prunable data
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	res, err := c.ap.bus.PrunableData(ctx)
+	cancel()
+
+	// return early if we couldn't fetch the prunable data
+	if err != nil {
+		c.logger.Error(err)
+		return
+	}
+
+	// return early if there's no prunable data
+	if res.TotalPrunable == 0 {
+		c.logger.Info("no contracts to prune")
+		return
+	}
+
+	// prune every contract individually, one at a time and for a maximum
+	// duration of 'timeoutPruneContract' to limit the amount of time we lock
+	// the contract as contracts on old hosts can take a long time to prune
+	var total uint64
+	wp.withWorker(func(w Worker) {
+		for _, contract := range res.Contracts {
+			c.logger.Debugf("pruning %d bytes from contract %v", contract.Prunable, contract.ID)
+			pruned, err := c.pruneContract(w, contract.ID)
+			if err != nil {
+				if isErr(err, errConnectionRefused) ||
```

Comment:
This seems quite hard to keep up-to-date. How about the opposite approach? When in doubt we continue pruning and we only check for errors which we can't recover from.

Reply:
I'm not sure I can predict errors which we can't recover from other than the `bus` not being accessible, so I decided to remove the `return` all together.